--- a/kiosk-shell/kiosk-shell.c	2022-09-23 01:16:14.000000000 +0900
+++ b/kiosk-shell/kiosk-shell.c	2022-10-15 07:53:07.911312903 +0900
@@ -413,6 +413,7 @@
 	weston_layer_entry_insert(&shsurf->shell->normal_layer.view_list,
 				  &shsurf->view->layer_link);
 	weston_view_geometry_dirty(shsurf->view);
+	weston_view_update_transform(shsurf->view);
 	weston_surface_damage(shsurf->view->surface);
 }
 
